server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1539,6 +1557,126 @@ public RepositoriesService getRepositoriesService() {
+            return updateWithRoutingTable(shardChanges, oldSnapshot, newRoutingTable);
+        }
+
+        @Override
+        protected void onChanged(ShardId shardId) {
+            shardChanges.add(shardId);
+        }
+    }
+
+    private static SnapshotsInProgress updateWithRoutingTable(Set<ShardId> shardIds, SnapshotsInProgress oldSnapshot,
+        RoutingTable newRoutingTable) {
+        if (oldSnapshot == null || shardIds.isEmpty()) {
+            return oldSnapshot;
+        }
+        List<SnapshotsInProgress.Entry> entries = new ArrayList<>();
+        boolean snapshotsInProgressChanged = false;
+        for (SnapshotsInProgress.Entry entry : oldSnapshot.entries()) {
+            ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shardsBuilder = null;
+            for (ShardId shardId : shardIds) {
+                ImmutableOpenMap<ShardId, ShardSnapshotStatus> shards = entry.shards();
+                ShardSnapshotStatus currentStatus = shards.get(shardId);
+                if (currentStatus != null && currentStatus.state().completed() == false) {
+                    final ShardSnapshotStatus newStatus = Optional
+                        .ofNullable(newRoutingTable.shardRoutingTableOrNull(shardId))
+                        .map(IndexShardRoutingTable::primaryShard)
+                        .map(
+                            primaryShardRouting -> determineShardSnapshotStatus(currentStatus, primaryShardRouting)
+                        )
+                        .orElse(failedStatus(null, "missing shard"));
+                    if (newStatus != currentStatus) {
+                        if (shardsBuilder == null) {
+                            shardsBuilder = ImmutableOpenMap.builder(shards);
+                        }
+                        shardsBuilder.put(shardId, newStatus);
+                    }
+                }
+            }
+            if (shardsBuilder == null) {
+                entries.add(entry);
+            } else {
+                snapshotsInProgressChanged = true;
+                ImmutableOpenMap<ShardId, ShardSnapshotStatus> shards = shardsBuilder.build();
+                entries.add(
+                    new SnapshotsInProgress.Entry(
+                        entry,
+                        completed(shards.values()) ? State.SUCCESS : entry.state(),
+                        shards
+                    )
+                );
+            }
+        }
+        if (snapshotsInProgressChanged) {
+            return new SnapshotsInProgress(entries);
+        }
+        return oldSnapshot;
+    }
+
+    private static ShardSnapshotStatus determineShardSnapshotStatus(final ShardSnapshotStatus currentStatus,
+        final ShardRouting primaryShardRouting) {
+        final State currentState = currentStatus.state();
+        final ShardSnapshotStatus newStatus;
+        if (primaryShardRouting == null) {
```

Comment:
can this ever be null here?

Reply:
Oh no, my bad for not cleaning that up, the caller ensures this is never `null` :)